# SI 330 - Homework #1: Data Manipulation

## Background

This homework assignment focuses on the analysis of historical data from the Olympic games.  The description of the data includes the following:
> This is a historical dataset on the modern Olympic Games, including all the Games from Athens 1896 to Rio 2016. I scraped this data from www.sports-reference.com in May 2018.

Your main task in this assignment is to explore the data *using the data
manipulation methods we covered in class* as well as those in the assigned readings.  You may need to consult pandas documentation, Stack Overflow, or other online resources.  

** You should also feel free to ask questions on the [class Slack homework channel] (https://si330wn2019.slack.com/messages/CFA5AJPCL/)! **

A total of 50 points is available in this homework assignment, consisting of:
- 40 points for completing the specific homework questions. More comprehensive 
answers will tend to gain more points.
- 5 points for the overall quality of spelling, grammar, punctuation, and style of written responses.  (see https://faculty.washington.edu/heagerty/Courses/b572/public/StrunkWhite.pdf for a detailed specifications).
- 5 points for creating code that conforms to [PEP 8](https://www.python.org/dev/peps/pep-0008/) guidelines.  You should review those guidelines before proceding with the assignment.


## Answer the questions below. 
For each question, you should
1. Write code using Python and pandas that can help you answer the following questions, and
2. Explain your answers in plain English. You should use complete sentences that would be understood by an educated professional who is not necessarily a data scientist (like a product manager).

### <font color="magenta"> Q1 (6 points): Explore and Describe the dataset. 
- How many rows and columns are there in the data frame? Please use the appropriate dataframe property.
- Provide summary statistics (i.e. use the .describe() function) for age, height, and weight. Please present this information in a single dataframe.

In [2]:
import pandas as pd
import numpy as np
from math import *

In [6]:
# read and open athlete_events.csv
olympics = pd.read_csv('data/athlete_events.csv')
# finds dimensions of data file
olympics.shape

There are 15 columns and 271116 rows. 

In [120]:
o1 = olympics[['Age', 'Height', 'Weight']].copy()

In [121]:
o1.describe()

Age         Height         Weight
count  261642.000000  210945.000000  208241.000000
mean       25.556898     175.338970      70.702393
std         6.393561      10.518462      14.348020
min        10.000000     127.000000      25.000000
25%        21.000000     168.000000      60.000000
50%        24.000000     175.000000      70.000000
75%        28.000000     183.000000      79.000000
max        97.000000     226.000000     214.000000

I used the describe() function to find the summary statistic for age, height and weight.
There are 261642 entries for age. The mean age is 25.5. The min is 10.0 and the max is 97.0. 
There are 210945 entries for height. The mean height is 175.3. The min is 127.0 and the max is 226.0.
There are 208241 entries for weight. The mean weight is 70.7. The min is 25.0 and the max is 214.0.

### <font color="magenta"> Q2 (8 points): How many unique athletes are in the dataset? What proportion of unique athletes are female?
    - Ideally, your code should return a proportion without hardcoding.

In [129]:
number = olympics.Name
number.nunique()

134732

There are 134732 unique athletes in the dataset.

In [122]:
olympics[olympics.Sex=='F']

ID                                      Name Sex   Age  Height  \
4            5                  Christine Jacoba Aaftink   F  21.0   185.0   
5            5                  Christine Jacoba Aaftink   F  21.0   185.0   
6            5                  Christine Jacoba Aaftink   F  25.0   185.0   
7            5                  Christine Jacoba Aaftink   F  25.0   185.0   
8            5                  Christine Jacoba Aaftink   F  27.0   185.0   
9            5                  Christine Jacoba Aaftink   F  27.0   185.0   
26           8        Cornelia "Cor" Aalten (-Strannood)   F  18.0   168.0   
27           8        Cornelia "Cor" Aalten (-Strannood)   F  18.0   168.0   
32          13                        Minna Maarit Aalto   F  30.0   159.0   
33          13                        Minna Maarit Aalto   F  34.0   159.0   
34          14            Pirjo Hannele Aalto (Mattila-)   F  32.0   171.0   
79          21                 Ragnhild Margrethe Aamodt   F  27.0   163.0   
80          22                             Andreea Aanei   F  22.0   170.0   
87          26              Agnes Erika Aanonsen (-Eyde)   F  17.0   169.0   
91          29                      Willemien Aardenburg   F  22.0     NaN   
105         37                        Ann Kristin Aarnes   F  23.0   182.0   
135         49                              Moonika Aava   F  24.0   168.0   
136         49                              Moonika Aava   F  28.0   168.0   
161         65                         Patimat Abakarova   F  21.0   165.0   
163         67   Mariya Vasilyevna Abakumova (-Tarabina)   F  22.0   179.0   
164         67   Mariya Vasilyevna Abakumova (-Tarabina)   F  26.0   179.0   
167         69                         Tamara Abalde Daz   F  19.0   185.0   
176         74                          Mara Laura Abalo   F  30.0   182.0   
193         88  Nunu Dzhansuhivna Abashydze (-Myslaieva)   F  25.0   168.0   
195         90                 Tamila Rashidovna Abasova   F  21.0   163.0   
196         90                 Tamila Rashidovna Abasova   F  21.0   163.0   
202         95                         Georgia Abatzidou   F  35.0   155.0   
252        126                            Forough Abbasi   F  20.0   164.0   
256        128                       Soulmaz Abbasi Azad   F  28.0   173.0   
259        131                             Tran Abbasova   F  33.0   159.0   
...        ...                                       ...  ..   ...     ...   
271008  135520                               Julia Zwehl   F  24.0   167.0   
271009  135520                               Julia Zwehl   F  28.0   167.0   
271027  135531      Sabine Beate Zwiener (-Jauchstetter)   F  24.0   172.0   
271029  135533     Petronella Veronica Maria "Nel" Zwier   F  23.0   174.0   
271032  135535                 Claudia Antoinette Zwiers   F  22.0   181.0   
271033  135535                 Claudia Antoinette Zwiers   F  30.0   181.0   
271050  135546                    Sandra Kornelia Zwolle   F  26.0   174.0   
271051  135546                    Sandra Kornelia Zwolle   F  26.0   174.0   
271052  135547             Viktoriya Viktorovna Zyabkina   F  19.0   174.0   
271053  135547             Viktoriya Viktorovna Zyabkina   F  23.0   174.0   
271054  135547             Viktoriya Viktorovna Zyabkina   F  23.0   174.0   
271055  135547             Viktoriya Viktorovna Zyabkina   F  23.0   174.0   
271057  135549             Nataliya Vasilyevna Zyatikova   F  27.0   162.0   
271058  135549             Nataliya Vasilyevna Zyatikova   F  27.0   162.0   
271059  135549             Nataliya Vasilyevna Zyatikova   F  27.0   162.0   
271060  135549             Nataliya Vasilyevna Zyatikova   F  27.0   162.0   
271061  135549             Nataliya Vasilyevna Zyatikova   F  27.0   162.0   
271062  135550                 Vera Vasilyevna Zyatikova   F  27.0   161.0   
271063  135550                 Vera Vasilyevna Zyatikova   F  27.0   161.0   
271064  135550                 Ver

In [130]:
sex = olympics.groupby('Sex').size()
sex

Sex
F     74522
M    196594
dtype: int64

(Use this space to explain your answers)

In [131]:
sex.loc['F']/(sex.loc['F']+sex.loc['M'])

0.2748712728131132

Around 0.275 of the unique athletes are females.

### <font color="magenta"> Q3 (10 points): Looking at the time period from 1950 to today, which athelete competed in the most number of events? 
In which unique events did the athelete participate, and for what range of years? Which country did the athlete represent?

In [132]:
# filters athletes who participated from 1950 to today
athlete = olympics[olympics.Year>=1950].copy()
athlete.head()

ID                      Name Sex   Age  Height  Weight         Team  NOC  \
0   1                 A Dijiang   M  24.0   180.0    80.0        China  CHN   
1   2                  A Lamusi   M  23.0   170.0    60.0        China  CHN   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0  Netherlands  NED   
5   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0  Netherlands  NED   
6   5  Christine Jacoba Aaftink   F  25.0   185.0    82.0  Netherlands  NED   

         Games  Year  Season         City          Sport  \
0  1992 Summer  1992  Summer    Barcelona     Basketball   
1  2012 Summer  2012  Summer       London           Judo   
4  1988 Winter  1988  Winter      Calgary  Speed Skating   
5  1988 Winter  1988  Winter      Calgary  Speed Skating   
6  1992 Winter  1992  Winter  Albertville  Speed Skating   

                                Event Medal  
0         Basketball Men's Basketball   NaN  
1        Judo Men's Extra-Lightweight   NaN  
4    Speed Skating Women's 500 metres   NaN  
5  Speed Skating Women's 1,000 metres   NaN  
6    Speed Skating Women's 500 metres   NaN

In [11]:
# finds athlete id who competed in the most events
athlete.groupby('ID').size().idxmax()

89187

In [137]:
# finds athlete info with id 89187
most = athlete[athlete.ID==89187].copy()
most

ID         Name Sex   Age  Height  Weight   Team  NOC        Games  \
177408  89187  Takashi Ono   M  20.0   160.0    58.0  Japan  JPN  1952 Summer   
177409  89187  Takashi Ono   M  20.0   160.0    58.0  Japan  JPN  1952 Summer   
177410  89187  Takashi Ono   M  20.0   160.0    58.0  Japan  JPN  1952 Summer   
177411  89187  Takashi Ono   M  20.0   160.0    58.0  Japan  JPN  1952 Summer   
177412  89187  Takashi Ono   M  20.0   160.0    58.0  Japan  JPN  1952 Summer   
177413  89187  Takashi Ono   M  20.0   160.0    58.0  Japan  JPN  1952 Summer   
177414  89187  Takashi Ono   M  20.0   160.0    58.0  Japan  JPN  1952 Summer   
177415  89187  Takashi Ono   M  20.0   160.0    58.0  Japan  JPN  1952 Summer   
177416  89187  Takashi Ono   M  25.0   160.0    58.0  Japan  JPN  1956 Summer   
177417  89187  Takashi Ono   M  25.0   160.0    58.0  Japan  JPN  1956 Summer   
177418  89187  Takashi Ono   M  25.0   160.0    58.0  Japan  JPN  1956 Summer   
177419  89187  Takashi Ono   M  25.0   160.0    58.0  Japan  JPN  1956 Summer   
177420  89187  Takashi Ono   M  25.0   160.0    58.0  Japan  JPN  1956 Summer   
177421  89187  Takashi Ono   M  25.0   160.0    58.0  Japan  JPN  1956 Summer   
177422  89187  Takashi Ono   M  25.0   160.0    58.0  Japan  JPN  1956 Summer   
177423  89187  Takashi Ono   M  25.0   160.0    58.0  Japan  JPN  1956 Summer   
177424  89187  Takashi Ono   M  29.0   160.0    58.0  Japan  JPN  1960 Summer   
177425  89187  Takashi Ono   M  29.0   160.0    58.0  Japan  JPN  1960 Summer   
177426  89187  Takashi Ono   M  29.0   160.0    58.0  Japan  JPN  1960 Summer   
177427  89187  Takashi Ono   M  29.0   160.0    58.0  Japan  JPN  1960 Summer   
177428  89187  Takashi Ono   M  29.0   160.0    58.0  Japan  JPN  1960 Summer   
177429  89187  Takashi Ono   M  29.0   160.0    58.0  Japan  JPN  1960 Summer   
177430  89187  Takashi Ono   M  29.0   160.0    58.0  Japan  JPN  1960 Summer   
177431  89187  Takashi Ono   M  29.0   160.0    58.0  Japan  JPN  1960 Summer   
177432  89187  Takashi Ono   M  33.0   160.0    58.0  Japan  JPN  1964 Summer   
177433  89187  Takashi Ono   M  33.0   160.0    58.0  Japan  JPN  1964 Summer   
177434  89187  Takashi Ono   M  33.0   160.0    58.0  Japan  JPN  1964 Summer   
177435  89187  Takashi Ono   M  33.0   160.0    58.0  Japan  JPN  1964 Summer   
177436  89187  Takashi Ono   M  33.0   160.0    58.0  Japan  JPN  1964 Summer   
177437  89187  Takashi Ono   M  33.0   160.0    58.0  Japan  JPN  1964 Summer   
177438  89187  Takashi Ono   M  33.0   160.0    58.0  Japan  JPN  1964 Summer   
177439  89187  Takashi Ono   M  33.0   160.0    58.0  Japan  JPN  1964 Summer   

        Year  Season       City       Sport  \
177408  1952  Summer   Helsinki  Gymnastics   
177409  1952  Summer   Helsinki  Gymnastics   
177410  1952  Summer   Helsinki  Gymnastics   
177411  1952  Summer   Helsinki  Gymnastics   
177412  1952  Summer   Helsinki  Gymnastics   
177413  1952  Summer   Helsinki  Gymnastics   
177414  1952  Summer   Helsinki  Gymnastics   
177415  1952  Summer   Helsinki  Gymnastics   
177416  1956  Summer  Melbourne  Gymnastics   
177417  1956  Summer  Melbourne  Gymnastics   
177418  1956  Summer  Melbourne  Gymnastics   
177419  1956  Summer  Melbourne  Gymnastics   
177420  1956  Summer  Melbourne  Gymnastics   
177421  1956  Summer  Melbourne  Gymnastics   
177422  1956  Summer  Melbourne  Gymnastics   
177423  1956  Summer  Melbourne  Gymnastics   
177424  1960  Summer       Roma  Gymnastics   
177425  1960  Summer       Roma  Gymnastics   
177426  1960  Summer       Roma  Gymnastics   
177427  1960  Summer       Roma  Gymnastics   
177428  1960  Summer       Roma  Gymnastics   
177429  1960  Summer       Roma  Gymnastics   
177430  1960  Summer       Roma  Gymnastics   
177431  1960  Summer       Roma  Gymnastics   
177432  1964  Summer      Tokyo  Gymnastics   
177433  1964  Summer      Tokyo  Gymnastics   
177434  1964  Summer      Tokyo  Gymnastics   
177435  1964  Summer 

In [138]:
# used athlete id to find name and country of person who participated in the most events
most[['Name', 'Team']].head(1)

Name   Team
177408  Takashi Ono  Japan

From 1950 to today, Takashi Ono from Japan participated in the most Olympic events

In [139]:
# Finds the events Takashi Ono participated in
most['Event']

177408    Gymnastics Men's Individual All-Around
177409          Gymnastics Men's Team All-Around
177410           Gymnastics Men's Floor Exercise
177411              Gymnastics Men's Horse Vault
177412            Gymnastics Men's Parallel Bars
177413           Gymnastics Men's Horizontal Bar
177414                    Gymnastics Men's Rings
177415          Gymnastics Men's Pommelled Horse
177416    Gymnastics Men's Individual All-Around
177417          Gymnastics Men's Team All-Around
177418           Gymnastics Men's Floor Exercise
177419              Gymnastics Men's Horse Vault
177420            Gymnastics Men's Parallel Bars
177421           Gymnastics Men's Horizontal Bar
177422                    Gymnastics Men's Rings
177423          Gymnastics Men's Pommelled Horse
177424    Gymnastics Men's Individual All-Around
177425          Gymnastics Men's Team All-Around
177426           Gymnastics Men's Floor Exercise
177427              Gymnastics Men's Horse Vault
177428            Gy

See above for list of events Takashi Ono participated in

In [27]:
# Finds the last year Takashi Ono participated in the olympics
most.loc[most['Year'].idxmax()]

ID                                         89187
Name                                 Takashi Ono
Sex                                            M
Age                                           33
Height                                       160
Weight                                        58
Team                                       Japan
NOC                                          JPN
Games                                1964 Summer
Year                                        1964
Season                                    Summer
City                                       Tokyo
Sport                                 Gymnastics
Event     Gymnastics Men's Individual All-Around
Medal                                        NaN
Name: 177432, dtype: object

The last year Takashi Ono participated in the Olympics was 1964.

In [140]:
# Finds the first year Takashi Ono participated in the olympics
most.loc[most['Year'].idxmin()]

ID                                         89187
Name                                 Takashi Ono
Sex                                            M
Age                                           20
Height                                       160
Weight                                        58
Team                                       Japan
NOC                                          JPN
Games                                1952 Summer
Year                                        1952
Season                                    Summer
City                                    Helsinki
Sport                                 Gymnastics
Event     Gymnastics Men's Individual All-Around
Medal                                        NaN
Name: 177408, dtype: object

The first year Takashi Ono participated in the Olympics was 1952.

Takashi Ono from Japan competed from 1952 to 1964.

### <font color="magenta"> Q4 (7 points): Which sport has the lowest median athlete age? Is there a tie?

In [143]:
# makes copy of dataframe sorting by age and sport
o2 = olympics[['Age', 'Sport']].copy()
o2.head()

Age          Sport
0  24.0     Basketball
1  23.0           Judo
2  24.0       Football
3  34.0     Tug-Of-War
4  21.0  Speed Skating

In [150]:
# finds the median age for each sport and lists them in ascending order
o2.groupby('Sport').median().sort_values(by='Age', ascending = True).head()

Age
Sport                      
Rhythmic Gymnastics    18.0
Swimming               20.0
Synchronized Swimming  22.0
Figure Skating         22.0
Diving                 22.0

Rhythmic gymnastic has the lowest median age of 18.0. There is no tie.

### <font color="magenta"> Q5 (8 points): In which year did Canada win the most medals?

In [117]:
# filters only data with Canada
o3 = olympics[olympics.Team=='Canada']
o3.head()

ID                       Name Sex   Age  Height  Weight    Team  NOC  \
269  140  William "Bill" Abbott Jr.   M  42.0   172.0    80.0  Canada  CAN   
270  140  William "Bill" Abbott Jr.   M  46.0   172.0    80.0  Canada  CAN   
279  146              Jeremy Abbott   M  19.0   179.0    71.0  Canada  CAN   
280  147              Joanne Abbott   F  41.0   160.0    57.0  Canada  CAN   
281  148     Kathryn "Katie" Abbott   F  21.0   164.0    63.0  Canada  CAN   

           Games  Year  Season      City     Sport  \
269  1996 Summer  1996  Summer   Atlanta   Sailing   
270  2000 Summer  2000  Summer    Sydney   Sailing   
279  1976 Summer  1976  Summer  Montreal  Canoeing   
280  1996 Summer  1996  Summer   Atlanta   Sailing   
281  2008 Summer  2008  Summer   Beijing   Sailing   

                                             Event Medal  
269            Sailing Mixed Three Person Keelboat   NaN  
270            Sailing Mixed Three Person Keelboat   NaN  
279  Canoeing Men's Canadian Doubles, 1,000 metres   NaN  
280            Sailing Mixed Three Person Keelboat   NaN  
281          Sailing Women's Three Person Keelboat   NaN

In [160]:
# sorts by data by year and medal
new_o3 = o3[['Year', 'Medal']].sort_values(by='Year', ascending=True).copy()
new_o3.head()

Year   Medal
83481   1900     NaN
178304  1900  Bronze
178306  1900     NaN
83480   1900     NaN
83497   1900     NaN

In [159]:
# finds the year with the most medal counts by sorting by descending order
new_o3.groupby('Year').count().sort_values(by='Medal', ascending=False).head(1)

Medal
Year       
1984     91

Canada won the most medals in 1981. Canada won 91 medals.

### <font color="magenta"> Q6 (11 points): In which sports in the 2014 Winter Olympics did female medalists win more medal points than male medalists?

In [233]:
o6 = olympics.copy()
# filters data in only the 2014 olympics
winter = o6[o6.Year==2014]
# sets and counts point values for medals
medal = winter[(winter.Medal=='Gold') |(winter.Medal=='Silver') | (winter.Medal=='Bronze')]
for i in medal['Medal']:
    if i == 'Gold':
        medal['Points'] = 3
    elif i == 'Silver':
        medal['Points'] = 2
    else:
        i == 'Bronze'
        medal['Points'] = 1

C:\Users\jodiy\Anaconda3.1\lib\site-packages\ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  del sys.path[0]
C:\Users\jodiy\Anaconda3.1\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\jodiy\Anaconda3.1\lib\site-packages\ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/s

In [234]:
# separates data by sex
male = medal[medal.Sex=='M']
female = medal[medal.Sex=='F']

In [227]:
# finds points for male medalists for each sport
male.groupby('Sport').Points.sum()

Sport
Alpine Skiing                 48
Biathlon                      90
Bobsleigh                     54
Cross Country Skiing          90
Curling                       36
Figure Skating                69
Freestyle Skiing              45
Ice Hockey                   213
Luge                          54
Nordic Combined               54
Short Track Speed Skating     63
Skeleton                       9
Ski Jumping                   54
Snowboarding                  45
Speed Skating                 72
Name: Points, dtype: int64

In [235]:
# finds points for female medalists for each sport
female.groupby('Sport').Points.sum()

Sport
Alpine Skiing                 45
Biathlon                      90
Bobsleigh                     18
Cross Country Skiing          90
Curling                       36
Figure Skating                66
Freestyle Skiing              45
Ice Hockey                   177
Luge                          18
Short Track Speed Skating     66
Skeleton                       9
Ski Jumping                    9
Snowboarding                  45
Speed Skating                 81
Name: Points, dtype: int64

In the 2014 winter olympics, female medalists won more points than male medalists in short track speed skating (66 vs 63) and in speed skating (81 vs 72).

### QBonus 5 pt: For each year in which games were held, what proportion of the host country medalists were women? 
You will need to combine multiple datasets to complete the analysis.
<p>Suggested data: 
<p>- https://en.wikipedia.org/wiki/List_of_Olympic_Games_host_cities (note: we suggest you use pd.read_html)

In [ ]:
# put your code here

(Use this space to explain your answers)

## Please submit your completed notebook in .IPYNB and .HTML formats via Canvas